### DISCUSSION TITLE
Line number and source code (via pdb) for maui android

### DISCUSSION THREAD

**Discussion by tranb3r** [getsentry/sentry-dotnet#3509] [Q&A]
Hi,

I'm uploading pdb files to Sentry for a maui android app.
I've managed to get line number and source code in error reports.
However I'm having a couple of issues.

First, it does not seem to work when I enable LLVM or full Trimming.
```
<TrimMode>full</TrimMode>
<EnableLLVM>true</EnableLLVM>
```

Second, it seems to work only for the code of the app, not for external nuget packages, even if I manually upload the nuget's pdb.

Are these issues known limitations? Am I missing something?

Thanks

**Comment by jamescrosswell** [comment]
Hey @tranb3r, 

> First, it does not seem to work when I enable LLVM or full Trimming.

When enabling LLVM, I'm not seeing any debug information files that could be used to symbolicate native code... I've started this discussion in the dotnet/maui repo to see if I can get a bit more information on how we could do this:
- https://github.com/dotnet/maui/discussions/23932

If you can upvote that discussion, it might help us get an answer more quickly.

**Reply by tranb3r** [reply]
What about Trimming?

**Reply by jamescrosswell** [reply]
> What about Trimming?

I tested this with the Sentry.Samples.Maui app and I am seeing source context for errors even when publishing the app fully trimmed. 

At first, I was seeing "old source code" associated with the exception. I think that's because I was running this locally so the source hadn't yet been committed and pushed to github where it would be available via source link.

However, if I set `<EmbedAllSources>true</EmbedAllSources>` in the csproj file though, then exceptions are shown in fully trimmed apps, including the correct source context, even for source code that I haven't yet committed and pushed to a repository that Sentry has access to.

**Reply by tranb3r** [reply]
Indeed, it works with trimming on the sample app.
However, it only works when the error is happening in the top level assembly.
If you add a class library and trigger an error from the class library itself, it does not anymore (although it works when disabling trimming).

**Comment by jamescrosswell** [comment]
> Second, it seems to work only for the code of the app, not for external nuget packages, even if I manually upload the nuget's pdb.

I think this should work. The Sentry Docs talk about various [built in symbol servers that Sentry uses](https://docs.sentry.io/platforms/dotnet/data-management/debug-files/symbol-servers/)

@bitsandfoxes @bruno-garcia can you think of any caveats to this or why/when this might not work?

**Reply by tranb3r** [reply]
I've done more tests for nugets.

It seems to work sometimes, but it depends on the nuget:
- it works for `Flurl.Http 4.0.2` (symbols in snupkg, no PublicKeyToken)
- it works for `SecureHttpClient 2.2.6` (symbols in snupkg, no PublicKeyToken)
- it does not work for `Sentry 4.9.0` (although it also has symbols in snupkg)
- it does not work for `CommunityToolkit.Mvvm 8.2.2` (symbols alongside the dll)
- it does not work for `Microsoft.Extensions.Http 9.0.0`

Is it just a coincidence that both working nugets in this list do not have a PublicKeyToken ?

**Reply by jamescrosswell** [reply]
Hm, all of these seem to have source context using [Source link](https://github.com/dotnet/sourcelink/blob/main/docs/README.md#embedallsources) (although Community Toolkit doesn't have a Symbol Server):
- [Flurl.Http](https://nuget.info/packages/Flurl.Http/4.0.2): Valid with Symbol Server
- [SecureHttpClient](https://nuget.info/packages/SecureHttpClient/2.2.6): Valid with Symbol Server
- [Sentry](https://nuget.info/packages/Sentry/4.9.0): Valid with Symbol Server
- [CommunityToolkit.Mvvm](https://nuget.info/packages/CommunityToolkit.Mvvm/8.2.2): Valid
- [Microsoft.Extensions.Http](https://nuget.info/packages/Microsoft.Extensions.Http/9.0.0-preview.6.24327.7): Valid with Symbol Server

Maybe as @vaind suggested, we'd need a reproducible example for this one as there are quite a few moving parts with a build like this.

**Comment by vaind** [comment]
> EnableLLVM

Looks like Android now has AOT compilation. I don't think we support that yet as it requires custom code for other platforms

**Reply by vaind** [reply]
but it's fair to say that I may not remember this correctly, since there's so many ways and combinations how you can build stuff...

We've got some handling on-device with AssemblyReader but that doesn't seem to help in this case.

we'll definitely need a reproducible example and look into this closely, frame by frame, to determine if the info is available there and whether it allows us to do symbolication

**Comment by tranb3r** [comment]
@jamescrosswell 

I've created a sample for my issues, based on the sample app for maui.
It's available here: https://github.com/tranb3r/sentry-demos-dotnet-maui/tree/tests-pdb
I hope it will help you figure out a solution for my issues.

# Trimming issue

## No source context for code in class library, when trimming is enabled
Enable trimming in DotNetMaui.csproj: `<TrimMode>full</TrimMode>`
![test1b-class-library](https://github.com/user-attachments/assets/8dfc4fd1-2e57-4315-a91e-e0ce82c96557)
![test1b-with-trimming](https://github.com/user-attachments/assets/bfc31718-b4df-47cc-add2-9c9734f6c1b4)

## No issue for code in class library, without trimming
![test1b-class-library](https://github.com/user-attachments/assets/7f5a1538-5ff8-468b-aa85-d7cf6626f8c6)
![test1b-without-trimming](https://github.com/user-attachments/assets/7d02b0a5-30b6-4e69-b246-ad2c4b843141)

## No issue for code in main assembly, with or without trimming
![test1a-main-dll](https://github.com/user-attachments/assets/bfa135cb-5d37-46bf-aa1a-4d596379a4fb)
![test1a-with-trimming](https://github.com/user-attachments/assets/7c722424-1242-4a6b-ae03-4bb1645b0ec5)

# Nuget issue : source context is not always available, it depends on the nuget package
![test2-nuget](https://github.com/user-attachments/assets/5f74a77e-7f9e-48f8-ac6a-066b649461e9)
![test2-part1](https://github.com/user-attachments/assets/d8f9c370-e977-4da4-b4e5-c14779487832)
![test2-part2](https://github.com/user-attachments/assets/19a75dfd-25d1-422a-be12-84e8ae54ea5b)

# Remark about PDB upload
`EmbedAllSources=false` is not always working for me.
I need to disable SourceLink for PDB upload to work properly.
```
<EnableSourceLink>false</EnableSourceLink>
<PublishRepositoryUrl>false</PublishRepositoryUrl>
```
Make sure you get the following files uploaded or the tests won't work:
![pdb](https://github.com/user-attachments/assets/a650d3fb-da49-43b8-a8a0-096946827f2f)

# EnableLLVM
When enabling LLVM, I don't see the same files in the uploaded PDB page (2 files instead of 4).
Maybe this explains why I can't get Source Context in errors reported by Sentry?
![pdb-lvvm](https://github.com/user-attachments/assets/43c6fa4e-eb12-4643-a26a-7d9ff04e18d1)

**Reply by jamescrosswell** [reply]
> I've created a sample for my issues, based on the sample app for maui.
> It's available here: https://github.com/tranb3r/sentry-demos-dotnet-maui/tree/tests-pdb
> I hope it will help you figure out a solution for my issues.

Thanks @tranb3r ... interestingly, I'm having even less success with your sample project than you are. I'm not able to see source context in Sentry under any circumstances, even including the following in the `*.csproj` file:

```xml
	<PropertyGroup>
		<EmbedAllSources>true</EmbedAllSources>
		<EnableSourceLink>false</EnableSourceLink>
		<PublishRepositoryUrl>false</PublishRepositoryUrl>
		<DebugType>portable</DebugType>
		<DebugSymbols>true</DebugSymbols>
		<TrimMode>partial</TrimMode>
		<Optimize>false</Optimize>
	</PropertyGroup>
```

The symbols and sources are being uploaded correctly to Sentry:
<img width="1165" alt="image" src="https://github.com/user-attachments/assets/64589617-7a17-46fa-8913-185a15577e12">

I suspected the stack traces themselves didn't include enough information to be symbolicated so tweaked the program to output debug information using the `FileDiagnosticLogger` (to see what stack trace was being sent with each of the envelopes for the crash reports). 

#### Optimization == true

```
"stacktrace":{"frames":[{"function":"void MainPage.OnCapturedExceptionClicked(object sender, EventArgs e)","in_app":true,"package":"DotNetMaui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instruction_addr":"0x0","addr_mode":"rel:0","function_id":"0xe"}]}
```

<img width="1572" alt="image" src="https://github.com/user-attachments/assets/b8a7c4ea-1720-4721-9599-585c1c76024b">


#### Optimization == false

```
"stacktrace":{"frames":[{"function":"void MainPage.OnCapturedExceptionClicked(object sender, EventArgs e)","in_app":true,"package":"DotNetMaui, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null","instruction_addr":"0x2","addr_mode":"rel:0","function_id":"0xe"}]},
```

<img width="1579" alt="image" src="https://github.com/user-attachments/assets/5da35b82-1f02-4900-963a-4f44c91bf7ba">

I think following the cookie trail from here means leaving the .NET SDK and Sentry-CLI and checking what happens on the Sentry server when it receives these envelopes and tries to symbolicate. Why isn't it able to either locate the relevant source context and/or symbolicate these stack traces correctly?

I've never setup a Sentry server locally for debugging though... so would need to get a hand from someone else here I think.

**Reply by tranb3r** [reply]
> Thanks @tranb3r ... interestingly, I'm having even less success with your sample project than you are. I'm not able to see source context in Sentry under any circumstances

@jamescrosswell 
I was also having a lot of difficulties to get reliable results, until I figured out a few things.

First, in csproj, you have to remove `<EmbedAllSources>true</EmbedAllSources>`.
Keep only the lines disabling `EnableSourceLink` and `PublishRepositoryUrl`.

Second, Sentry's server seem to cache some symbol files.
Even when they do not appear on the web page, the server seem to keep using some of them. 
So here is what I do between 2 tests, in order to make sure I'm uploading new files and Sentry is not using old files that are hidden:
- manually delete every uploaded PDB files, on the Sentry web page.
- make sure you do one change in each assembly, before building again.
- manually upload files with sentry-cli (not the msbuild task), with debug verbosity.
- check sentry-cli log, and make sure it has uploaded 2 files per assembly, with sources (not skipping them).
- finally, check sentry web page, and make sure 2 brand new files have been uploaded for each assembly (`SourceBundle` and `Portable PDB debug companion`).

**Reply by jamescrosswell** [reply]
Hm, that sounds like a lot of secret handshakes and like the process is far too fragile to be usable/useful for most SDK users. 

> manually upload files with sentry-cli (not the msbuild task), with debug verbosity.

@tranb3r what command are you running, exactly, to perform the uploads? Is it the same as what the MSBuild task runs (only with debug verbosity) or are you running a different command for some reason?

**Reply by jamescrosswell** [reply]
OK, I finally managed to get the source context to show for a Mac Catalyst app by setting:
```xml
  <MtouchLink>None</MtouchLink>
```

... which, obviously, you woudn't want to do in a production application since it makes the binary huge.

Not to mention:
```xml
  <TrimMode>partial</TrimMode>
  <Optimize>false</Optimize>
```

All in all, I had the feeling I was making random changes and getting random results. Things get uploaded to Sentry. Sometimes it works and sometimes it doesn't but it's hard to know why, either way. It's very much a black box. 

It would be great if there was some way to get "logs" from Sentry detailing how it succeeds or fails to resolve symbols and source context for stack traces.

**Reply by jamescrosswell** [reply]
@tranb3r I've created a features request in the Sentry repo for this:
- https://github.com/getsentry/sentry/issues/75705

**Reply by tranb3r** [reply]
> that sounds like a lot of secret handshakes and like the process is far too fragile to be usable/useful for most SDK users.

It's only necessary when iterating on a developer machine, like we've done.
In a CI context, it would be much simpler, since the assemblies version would be increased (so there would be no caching).

> @tranb3r what command are you running, exactly, to perform the uploads? Is it the same as what the MSBuild task runs (only with debug verbosity) or are you running a different command for some reason?

The reason why I've chosen to perform the uploads with scentry-cli instead of msbuild is because when you include the Sentry.Maui package in a class library, the Sentry target is executed both when building the class library and when building the application, resulting in the pdb upload being done twice. 
First upload would be before trimming, second upload after trimming... I imagine this could cause issues with Sentry server's caching.
So I've decided not to trust the msbuild target and run the command after the build. But the command should be the same.
`sentry-cli-Windows-x86_64.exe debug-files upload -o XXX -p YYY --include-sources --log-level debug [PATH]`

> All in all, I had the feeling I was making random changes and getting random results.  

I had this feeling too, until I figured out the steps I've described, that work for me.

> It's very much a black box. It would be great if there was some way to get "logs" from Sentry detailing how it succeeds or fails to resolve symbols and source context for stack traces.

This is a very good idea, I agree it would help a lot when trying to understand what's going on.

**Reply by bitsandfoxes** [reply]
Just to have mentioned it and to help with further debugging: The actual event-json contains further information regarding the symbolication status.
i.e. source link status, debug symbol ID and 
```
"data": {
    "symbolicator_status": "symbolicated"
},
```
